File: backend/app/models/timetable.py
# backend/app/models/timetable.py
from datetime import datetime
from typing import Optional
from sqlalchemy import String, Integer, Boolean, DateTime, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class TimetableRun(Base):
    __tablename__ = "timetable_runs"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(200))
    status: Mapped[str] = mapped_column(String(20), default="draft")
    # draft | under_review | approved | published
    semester_id: Mapped[int] = mapped_column(ForeignKey("semesters.id"))
    created_by: Mapped[int] = mapped_column(ForeignKey("users.id"))
    generated_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    semester: Mapped["Semester"] = relationship()
    creator: Mapped["User"] = relationship()
    faculties: Mapped[list["TimetableRunFaculty"]] = relationship(
        back_populates="run", cascade="all, delete-orphan"
    )
    buildings: Mapped[list["TimetableRunBuilding"]] = relationship(
        back_populates="run", cascade="all, delete-orphan"
    )
    entries: Mapped[list["TimetableEntry"]] = relationship(
        back_populates="run", cascade="all, delete-orphan"
    )
    conflicts: Mapped[list["TimetableConflict"]] = relationship(
        back_populates="run", cascade="all, delete-orphan"
    )
    jobs: Mapped[list["GenerationJob"]] = relationship(
        back_populates="run", cascade="all, delete-orphan"
    )
    approvals: Mapped[list["FacultyHeadApproval"]] = relationship(
        back_populates="run", cascade="all, delete-orphan"
    )


class TimetableRunFaculty(Base):
    __tablename__ = "timetable_run_faculties"

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("timetable_runs.id"))
    faculty_id: Mapped[int] = mapped_column(ForeignKey("faculties.id"))

    run: Mapped["TimetableRun"] = relationship(back_populates="faculties")
    faculty: Mapped["Faculty"] = relationship()


class TimetableRunBuilding(Base):
    __tablename__ = "timetable_run_buildings"

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("timetable_runs.id"))
    building_id: Mapped[int] = mapped_column(ForeignKey("buildings.id"))

    run: Mapped["TimetableRun"] = relationship(back_populates="buildings")
    building: Mapped["Building"] = relationship()


class TimetableEntry(Base):
    __tablename__ = "timetable_entries"

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("timetable_runs.id"))
    course_id: Mapped[int] = mapped_column(ForeignKey("courses.id"))
    lecturer_id: Mapped[int] = mapped_column(ForeignKey("lecturers.id"))
    room_id: Mapped[int] = mapped_column(ForeignKey("rooms.id"))
    time_slot_id: Mapped[int] = mapped_column(ForeignKey("time_slots.id"))
    group_id: Mapped[Optional[int]] = mapped_column(ForeignKey("class_groups.id"), nullable=True)
    week_pattern: Mapped[str] = mapped_column(String(20), default="every_week")
    rotation_sequence: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_overcapacity: Mapped[bool] = mapped_column(Boolean, default=False)
    is_merged: Mapped[bool] = mapped_column(Boolean, default=False)

    run: Mapped["TimetableRun"] = relationship(back_populates="entries")
    course: Mapped["Course"] = relationship()
    lecturer: Mapped["Lecturer"] = relationship()
    room: Mapped["Room"] = relationship()
    time_slot: Mapped["TimeSlot"] = relationship()
    group: Mapped[Optional["ClassGroup"]] = relationship()
    entry_classes: Mapped[list["TimetableEntryClass"]] = relationship(
        back_populates="entry", cascade="all, delete-orphan"
    )


class TimetableEntryClass(Base):
    __tablename__ = "timetable_entry_classes"

    id: Mapped[int] = mapped_column(primary_key=True)
    entry_id: Mapped[int] = mapped_column(ForeignKey("timetable_entries.id"))
    class_id: Mapped[int] = mapped_column(ForeignKey("classes.id"))

    entry: Mapped["TimetableEntry"] = relationship(back_populates="entry_classes")
    student_class: Mapped["Class"] = relationship()


class TimetableConflict(Base):
    __tablename__ = "timetable_conflicts"

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("timetable_runs.id"))
    conflict_type: Mapped[str] = mapped_column(String(50))
    course_id: Mapped[Optional[int]] = mapped_column(ForeignKey("courses.id"), nullable=True)
    class_id: Mapped[Optional[int]] = mapped_column(ForeignKey("classes.id"), nullable=True)
    details: Mapped[str] = mapped_column(Text, default="{}")
    resolved: Mapped[bool] = mapped_column(Boolean, default=False)
    resolution: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)

    run: Mapped["TimetableRun"] = relationship(back_populates="conflicts")
    course: Mapped[Optional["Course"]] = relationship()
    student_class: Mapped[Optional["Class"]] = relationship()


class GenerationJob(Base):
    __tablename__ = "generation_jobs"

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("timetable_runs.id"))
    status: Mapped[str] = mapped_column(String(20), default="pending")
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    started_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    run: Mapped["TimetableRun"] = relationship(back_populates="jobs")


class FacultyHeadApproval(Base):
    __tablename__ = "faculty_head_approvals"

    id: Mapped[int] = mapped_column(primary_key=True)
    run_id: Mapped[int] = mapped_column(ForeignKey("timetable_runs.id"))
    faculty_id: Mapped[int] = mapped_column(ForeignKey("faculties.id"))
    faculty_head_id: Mapped[int] = mapped_column(ForeignKey("users.id"))
    status: Mapped[str] = mapped_column(String(20), default="pending")
    # pending | approved | rejected
    comment: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    decided_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    run: Mapped["TimetableRun"] = relationship(back_populates="approvals")
    faculty: Mapped["Faculty"] = relationship()
    faculty_head: Mapped["User"] = relationship()


class Notification(Base):
    __tablename__ = "notifications"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"))
    message: Mapped[str] = mapped_column(String(500))
    notification_type: Mapped[str] = mapped_column(String(50))
    # timetable_published | slot_changed | conflict_flagged | generation_complete
    # timetable_review | timetable_approved | timetable_rejected
    is_read: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    user: Mapped["User"] = relationship()

File: frontend/lib/core/models/timetable.dart
class TimetableModel {
  final int id;
  final String status;
  final int semesterId;
  final int departmentId;
  final String? generatedAt;
  final String createdAt;

  const TimetableModel({
    required this.id, required this.status, required this.semesterId,
    required this.departmentId, this.generatedAt, required this.createdAt,
  });

  factory TimetableModel.fromJson(Map<String, dynamic> json) => TimetableModel(
        id: json['id'] as int,
        status: json['status'] as String,
        semesterId: json['semester_id'] as int,
        departmentId: json['department_id'] as int,
        generatedAt: json['generated_at'] as String?,
        createdAt: json['created_at'] as String,
      );

  bool get isPublished => status == 'published';
  bool get isDraft => status == 'draft';
}

class TimetableEntry {
  final int id;
  final int timetableId;
  final int courseId;
  final int lecturerId;
  final int roomId;
  final int timeSlotId;
  final int? groupId;
  final String weekPattern;
  final String? rotationSequence;
  final bool isOvercapacity;
  final bool isMerged;
  final List<int> classIds;

  const TimetableEntry({
    required this.id, required this.timetableId, required this.courseId,
    required this.lecturerId, required this.roomId, required this.timeSlotId,
    this.groupId, required this.weekPattern, this.rotationSequence,
    required this.isOvercapacity, required this.isMerged, required this.classIds,
  });

  factory TimetableEntry.fromJson(Map<String, dynamic> json) => TimetableEntry(
        id: json['id'] as int,
        timetableId: json['timetable_id'] as int,
        courseId: json['course_id'] as int,
        lecturerId: json['lecturer_id'] as int,
        roomId: json['room_id'] as int,
        timeSlotId: json['time_slot_id'] as int,
        groupId: json['group_id'] as int?,
        weekPattern: json['week_pattern'] as String,
        rotationSequence: json['rotation_sequence'] as String?,
        isOvercapacity: json['is_overcapacity'] as bool,
        isMerged: json['is_merged'] as bool,
        classIds: List<int>.from(json['class_ids'] as List),
      );
}

class TimetableConflict {
  final int id;
  final int timetableId;
  final String conflictType;
  final int? courseId;
  final int? classId;
  final String details;
  final bool resolved;
  final String? resolution;

  const TimetableConflict({
    required this.id, required this.timetableId, required this.conflictType,
    this.courseId, this.classId, required this.details,
    required this.resolved, this.resolution,
  });

  factory TimetableConflict.fromJson(Map<String, dynamic> json) => TimetableConflict(
        id: json['id'] as int,
        timetableId: json['timetable_id'] as int,
        conflictType: json['conflict_type'] as String,
        courseId: json['course_id'] as int?,
        classId: json['class_id'] as int?,
        details: json['details'] as String,
        resolved: json['resolved'] as bool,
        resolution: json['resolution'] as String?,
      );
}

class AppNotification {
  final int id;
  final String message;
  final String notificationType;
  final bool isRead;
  final String createdAt;

  const AppNotification({
    required this.id, required this.message, required this.notificationType,
    required this.isRead, required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) => AppNotification(
        id: json['id'] as int,
        message: json['message'] as String,
        notificationType: json['notification_type'] as String,
        isRead: json['is_read'] as bool,
        createdAt: json['created_at'] as String,
      );
}
